posthog/temporal/data_imports/pipelines/temporalio/source.py

```diff
@@ -0,0 +1,200 @@
+import asyncio
+import datetime
+from enum import Enum
+from queue import Queue
+import threading
+from typing import Any, Optional
+
+from posthog.temporal.data_imports.pipelines.pipeline.typings import SourceResponse
+from posthog.temporal.data_imports.pipelines.source import config
+from posthog.warehouse.types import IncrementalField, IncrementalFieldType
+from temporalio.client import Client
+from temporalio.service import TLSConfig, RPCError
+
+
+class TemporalIOResource(Enum):
+    Workflows = "workflows"
+    WorkflowHistories = "workflow_histories"
+
+
+ENDPOINTS = (TemporalIOResource.Workflows.value, TemporalIOResource.WorkflowHistories.value)
+INCREMENTAL_ENDPOINTS = (TemporalIOResource.Workflows.value, TemporalIOResource.WorkflowHistories.value)
+
+INCREMENTAL_FIELDS: dict[str, list[IncrementalField]] = {
+    TemporalIOResource.Workflows.value: [
+        {
+            "label": "CloseTime",
+            "type": IncrementalFieldType.DateTime,
+            "field": "close_time",
+            "field_type": IncrementalFieldType.DateTime,
+        }
+    ],
+    TemporalIOResource.WorkflowHistories.value: [
+        {
+            "label": "CloseTime",
+            "type": IncrementalFieldType.DateTime,
+            "field": "workflow_close_time",
+            "field_type": IncrementalFieldType.DateTime,
+        }
+    ],
+}
+
+
+@config.config
+class TemporalIOSourceConfig(config.Config):
+    host: str
+    port: int
+    namespace: str
+    server_client_root_ca: str
+    client_certificate: str
+    client_private_key: str
+
+
+def _async_iter_to_sync(async_iter):
+    q: Queue[Any] = Queue()
+    sentinel = object()
+
+    async def runner():
+        try:
+            async for item in async_iter:
+                q.put(item)
+        finally:
+            q.put(sentinel)
+
+    def run_event_loop():
+        asyncio.run(runner())
+
+    threading.Thread(target=run_event_loop, daemon=True).start()
+
+    while True:
+        item = q.get()
```

Comment:
If this gets blocked for too long, the queue could grow a lot and put memory pressure. Maybe we should consider a limit on the queue? Then, that would require calling `q.task_done()` after yielding too.

Reply:
Hmm, what does that look like here? i'm unsure on the impact of doing that?

Reply:
The `run_event_loop` thread runs to completion before main thread gets any chance to do `q.get()`. The queue gets populated with all available data. If there is a lot of data, that means a lot of memory used by the queue. Adding a queue size limit would block the `run_event_loop` thread ensuring the main thread has a chance to clear up some items, thus avoiding excessive memory usage at a time.

Reply:
> The run_event_loop thread runs to completion before main thread gets any chance to do q.get(). 

Im not sure if that's true. The message history endpoint is slow in terms of how many requests it has to make, our 5k batching still works as expected and the requesting seems to be paused in the background while we dealing with the first 5000 rows

Reply:
I think a handful of changes would be enough:
Adding a queue limit:
```
    q: Queue[Any] = Queue(maxsize=5000) # Or whatever matches the batch size + 1 (to account for the sentinel)
```

Calling `task_done`
```
        if item is sentinel:
            q.task_done()
            break
        yield item
        q.task_done()
```